[Auth] Add ability to login with a passkey.

notandyvee opened the pull request:
### Fix

This PR builds upon [1665](https://github.com/Automattic/simplenote-android/pull/1665). Once you have a passkey associated with your account you can verify it and login.

### Test

Note: This PR was only tested with google password manager.

- [ ] You must build the app in release and sign it with the release certificate.
- [ ] If you have not added a passkey to your account, please login using password first. Follow the instructions in [1665](https://github.com/Automattic/simplenote-android/pull/1665) in order to add a passkey.
- [ ] On the login screen: "Login" -> "Log In With Email"
- [ ] Enter the email you associated your passkey with.
- [ ] Click "Login With A Passkey"
- [ ] Follow the instructions and if successful you should be logged in.

### Note

I will address any issues when using passkeys other than Google Password Manager in a different PR.

notandyvee committed:
[Auth] Add ability to login with passkey

Note: Only tested with Google Password Manager.

dangermattic commented:
<!--
  0 Errors
  2 Warnings: <code>strings.xml</code> files...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><code>strings.xml</code> files should only be updated on release branches, when the translations are downloaded by our automation.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

notandyvee committed:
Merge branch 'trunk' into andy/passkey-login

notandyvee committed:
Merge branch 'andy/passkey-feature' into andy/passkey-login

notandyvee committed:
Fix failing tests.

notandyvee committed:
[Auth] Added basic support for discoverable passkeys.

But left it commented out since we don't yet know if we will
go with this. Also gated the passkey feature to only API 28 and
greater.

danilo04 commented on Simplenote/src/main/java/com/automattic/simplenote/authentication/MagicLinkableFragment.kt:
```diff
@@ -21,9 +21,10 @@ import com.simperium.android.ProgressDialogFragment
  */
 abstract class MagicLinkableFragment : Fragment() {
```
(⛏️) Why don't you call this `AbstractSignInFragment` since it is used for magic links and passcode?

notandyvee replied:
You are not wrong. Let me table this one. I don't want to make this update now because once I merge trunk, I will get a ton of merge conflicts. I can make the update in a separate PR if you don't mind.

notandyvee replied:
Actually, I ended up doing it. Didn't realize it is currently only being used by `SignInFragment`. The goal is to also use it for signup. So name it `BaseAuthenticationFragment`.

danilo04 commented on Simplenote/src/main/java/com/automattic/simplenote/authentication/SignInFragment.kt:
```diff
@@ -22,26 +30,49 @@ class SignInFragment: MagicLinkableFragment() {
+                    this.startActivity(intent)
+                    act.finish()
+                }
+            }
```
This logic for login with password should be outside the if since it should work with phones with Android 27 and older.

notandyvee replied:
Great catch! 🙏🏻

danilo04 commented on Simplenote/src/main/java/com/automattic/simplenote/authentication/MagicLinkableFragment.kt:
```diff
@@ -40,28 +41,36 @@ abstract class MagicLinkableFragment : Fragment() {
+        signupButton = view.findViewById(R.id.button)
+        signupButton?.text = actionButtonText()
 
         emailField?.let {
```
I would recommend using a name for the param since it will be used inside another lambda.

danilo04 reviewed (commented):
Hi @notandyvee, I tested the different scenarios and they are working as expected. I left a couple of comments, one important so the password functionality can work on devices older than Android 28.

Commit:
Merge branch 'trunk' into andy/passkey-login

Commit:
[Bug] Fix build issues and fix nullpointer crash.

Commit:
Renamed MagicLinkableFragment to BaseAuthenticationFragment

Commit:
Merge branch 'andy/passkey-feature' into andy/passkey-login

Commit:
Used name value for let block.

notandyvee commented:
Ready for re-review @danilo04

danilo04 reviewed (approved):
Thank you for the changes @notandyvee. This looks good :shipit:
